### ISSUE TITLE
Deadlock in task.upload_artifact()

### ISSUE THREAD

**Issue by bzamecnik** [allegroai/clearml-agent#73]
When I run my model training it freezes upon calling `task.upload_artifact()`. I'm not sure if it may be affected by `tensorflow`, `logging` or `multiprocessing`. It runs in the main process.

```
mp.set_start_method("forkserver", force=True)
logging.basicConfig(level=logging.INFO)
task = Task.init(
    project_name="Foo",
    task_name="Foo train",
    task_type=Task.TaskTypes.training,
    reuse_last_task_id=False,
)
task.connect(config)
task.upload_artifact("train.yaml", yaml.dumps(config))
```

When I kill the script (CTRL+C) it shows a traceback where its waiting on a lock `Artifacts._task_edit_lock`:

```
Traceback (most recent call last):
  File "foo_project/tools/train_model.py", line 201, in <module>
    args.delegate(parser, args)
  File "foo_project/tools/train_model.py", line 145, in train_model
    MODEL_TO_TRAIN_FUNC.get(kind, default_train_function)(config)
  File "/home/bzamecnik/foo_project/tools/model_train.py", line 81, in model_train
    task.upload_artifact("train.yaml", yaml.dumps(config))
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/task.py", line 1544, in upload_artifact
    auto_pickle=auto_pickle, preview=preview, wait_on_upload=wait_on_upload)
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/binding/artifacts.py", line 567, in upload_artifact
    self._add_artifact(artifact)
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/binding/artifacts.py", line 627, in _add_artifact
    self._task._add_artifacts(self._task_artifact_list)
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/backend_interface/task/task.py", line 1229, in _add_artifacts
    with self._edit_lock:
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/utilities/locks/utils.py", line 208, in __enter__
    return self.acquire()
  File "/home/bzamecnik/.virtualenvs/foo_env/lib/python3.6/site-packages/clearml/utilities/locks/utils.py", line 236, in acquire
    if not self._lock.acquire(block=timeout != 0, timeout=timeout):
  File "/home/bzamecnik/.virtualenvs/rossum3/lib/python3.6/site-packages/clearml/task.py", line 3171, in signal_handler
    return org_handler if not callable(org_handler) else org_handler(sig, frame)
```

System & packages:

- Ubuntu 18.04
- clearml==1.0.4
- tensorflow-gpu==2.3.3

I don't know what might have changed but it worked in TF 1.15 and even in 2.3.3 when I was testing that before. The only difference I'm aware it that the on-premise ClearML server was upgraded from some 0.17.x to 1.0.

I observed that if I add `print()` call around the `with self._task_edit_lock` blocks the deadlock does not occur. So it looks like some race condition depending on good timing.

Also it seems that avoiding the `mp.set_start_method("forkserver", force=True)` prevents the deadlock. IMHO the forkserver should be called first before importing anything else. In this case it was problematic. It seems the problem is caused by interaction of the `forkserver` and `clearml.utilities.mp.BackgroundMonitor`.

Hack: Adding some `time.sleep()` between `task.connect()` and `task.upload_artifact()` seems to avoid the race condition:

```
import time

task.connect(config)
time.sleep(2.0)
task.upload_artifact("train.yaml", yaml.dumps(config))
```

Minimal example for replication:

```
import multiprocessing as mp
if __name__ == '__main__':
    mp.set_start_method("forkserver")
import tensorflow.compat.v1 as tf
from clearml import Task

def main():
        task = Task.init(
        project_name="Debug",
        task_name="ClearML deadlock debug",
        task_type=Task.TaskTypes.training,
        reuse_last_task_id=False,
    )
    
    config = {'foo': 'bar'}
    
    task.connect(config)
    task.upload_artifact("train.txt", 'foo bar')
    
    print('Done')


if __name__ == '__main__':
    tf.disable_v2_behavior()
    main()
```

I found that the deadlock is caused by a combination of TF 2 `tf.disable_v2_behavior()` and `mp.set_start_method("forkserver")` (or `spawn`).

Note the number of threads:
- before `Task.init()`: 1
- after `Task.init()`: 4
- after `task.connect()`: 5
- after `task.upload_artifact()`: 4 (in case the deadlock does not happen

**Comment by bmartinn** [comment]
Thanks @bzamecnik 

> The only difference I'm aware it that the on-premise ClearML server was upgraded from some 0.17.x to 1.0.

Yes the major change in ClearML is that the reporting (and also artifact uploading) was moved from background thread to background process (to make sure that multi process tasks use a single upload process and hence are more efficient)
BTW to disable this behavior and revert to the old (background thread), you can add `sdk.development.report_use_subprocess: false` to your `~/clearml.conf` file.


> Also it seems that avoiding the mp.set_start_method("forkserver", force=True) prevents the deadlock. 

Yes, the change in thread->sub-process upsides is that `forkserver` is not a needed anymore. 

> In this case it was problematic. It seems the problem is caused by interaction of the forkserver and clearml.utilities.mp.BackgroundMonitor.

What do you mean by that? what exactly are you getting , an exception ?


Let me see if I can reproduce the issue, I suspect that the race here is that the forked process forks the state of the open files (i.e. the lock), and the subprocess will always think the lock is acquired, when in fact it was released.
wdyt?

**Comment by bzamecnik** [comment]
Hi @bmartinn! Thanks for the reply. Ahh, so a background process for reporting would likely explain that.

Setting `sdk.development.report_use_subprocess: false` avoid the problem. Thanks!

Anyway I'm working on transitioning our code to use the eager mode and avoid the v1 mode in Tensorflow, so we will be able to use the default setting of ClearML (background process for reporting).

The point of forkserver (in contrast to fork) should be to have a clean process to fork from, so that no bad state get copied to the children. Do you think it's worth investigating further when there's a workaround with using the thread for this combination? Or may it lead to some bug that can arise under other conditions?

**Comment by bmartinn** [comment]
> The point of forkserver (in contrast to fork) should be to have a clean process to fork from

Yes, I like this approach better as well, but we had some experience when the performance was slower compared to regular forks (only tested with pytorch DataLoader, so it might be a specific usage scenario that caused it, and has nothing to do with `clearml` but more to how the subprocesses are created and when)

> Do you think it's worth investigating further when there's a workaround with using the thread for this combination? Or may it lead to some bug that can arise under other conditions?

I think it just revealed a bug in the edit_lock, basically the Lock state should not be cloned, see logic [here](https://github.com/allegroai/clearml/blob/d585535dcdb8e95398be0ff5b1831159832addc8/clearml/backend_interface/task/task.py#L2157)
It seems to be using the "main" process FileLock, which should check the if it needs to recreate itself if used on a child process. I might have an idea on how to fix it. Is it easy to reproduce ?

**Comment by bzamecnik** [comment]
Yes, it was easy and consistent. You can use the script above with the dependencies stated there. Possibly within a tensorflow docker image or something. Good luck with fixing the locks.

**Comment by bzamecnik** [comment]
Now I have realized I likely posted this issue in the wrong project. It should not be clearml-agent, but clearml (client).

**Comment by bmartinn** [comment]
> Now I have realized I likely posted this issue in the wrong project. It should not be clearml-agent, but clearml (client).

Correct, no worries :) when closed I'll see how we can move it.

> You can use the script above with the dependencies stated there. Possibly within a tensorflow docker image or something. Good luck with fixing the locks.

Okay let's hope I can also reproduce it in a consistent way (this is the challenge with race conditions ;)

**Comment by bmartinn** [comment]
@bzamecnik good news, I was able to reproduce the issue.
(1) post v1.0 there is not need to call `mp.set_start_method("forkserver")` before the `clearml` import.
(2) The only way I was able to reproduce the issue was when I disabled `tf.disable_v2_behavior()` , not really sure why...

Regrading the actual bug, this seems to be a Python issue, https://bugs.python.org/issue6721
Essentially the problem is that when forking a process when one of the threads has an acquired process RLock, the new forked process lock internal state is not cleared.
This seems to happen only when using `forkserver` as the multiprocessing context.
Anyhow, a fix will be committed soon :)
